bin/adhoc-scripts/createPileupObjects.py

```diff
@@ -0,0 +1,179 @@
+    from WMCore.Services.pycurl_manager import RequestHandler
+except ImportError:
+    print("WMCore/RequestHandler environment not set. Faking it")
+    RequestHandler = None
+try:
+    from WMCore.Services.ReqMgrAux.ReqMgrAux import ReqMgrAux
+except ImportError:
+    print("WMCore/ReqMgrAux environment not found. Faking it")
+    ReqMgrAux = None
+
+
+class OptionParser():
+    """Class to parse the command line arguments"""
+
+    def __init__(self):
+        "User based option parser"
+        self.parser = argparse.ArgumentParser(prog='PROG')
+        self.parser.add_argument("--fin", action="store",
+                                 dest="fin", default="", help="Input file")
+        self.parser.add_argument("--fout", action="store",
+                                 dest="fout", default="",
+                                 help="Output json file with all the pileup configurations created")
+        self.parser.add_argument("--verbose", action="store",
+                                 dest="verbose", default=0, help="verbosity level")
+        self.parser.add_argument("--url", action="store",
+                                 dest="url", default="https://cmsweb-testbed.cern.ch",
+                                 help="Server url to fetch and create pileup configurations")
+        self.parser.add_argument("--inject", action="store_true",
+                                 dest="inject", default=False,
+                                 help="Inject pileup documents for real")
+
+
+def getPUSchema(pileupName, pileupDocs, logger):
+    """
+    Get a minimalistic pileup data schema, or return an
+    existent pileup object.
+
+    :param pileupName: string with the pileup name
+    :param pileupDocs: list of pileup documents
+    :return: a dictionary with the pileup configuration
+    """
+    for doc in pileupDocs:
+        if pileupName == doc["pileupName"]:
+            logger.warning("Reusing pileup configuration for pileup: %s", pileupName)
+            return doc
+    pileupDoc = {
+        'pileupName': "",
+        'pileupType': "",
+        'expectedRSEs': [],
+        'campaigns': [],
+        'active': False}
+    return pileupDoc
+
+
+def parseCampaigns(campDocs, logger):
+    """
+    Parse campaigns from WMCore and convert them into
+    Pileup documents, where:
+     * campaigns: is a union of all campaigns using a given pileup.
+     * expectedRSEs: is a union of all RSEs across different campaigns.
+    """
+    pileupDocs = []
+    for camp in campDocs:
+        # for each secondary dataset, create one pileup document
+        for puName, puRSEs in camp.get("Secondaries", {}).items():
+            puDoc = getPUSchema(puName, pileupDocs, logger)
+            puType = "premix" if puName.split("/")[-1] == "PREMIX" else "classic"
+            puDoc["pileupName"] = puName
+            puDoc["pileupType"] = puType
+            puDoc["expectedRSEs"].extend(puRSEs)
+            puDoc["campaigns"].append(camp["CampaignName"])
```

Comment:
here you assume that "CampaignName" exists in camp dictionary which is error prone. Instead, I suggest to check if such key exists in camp dictionary and its value is valid string before you'll append it to pileup doc.

Reply:
I agree with this statement when we are parsing data that is unknown or dynamic, but here we are fetching the WMCore campaigns that are well defined and stored in central CouchDB.

Reply:
According to my code reading the code is not fetch docs directly from CouchDB, instead it relies on ReqMgrAux class which by itself relies on MemoryCacheStruct which by itself is populated by HTTP calls to CouchDB. Without any knowledge how cache works I have no strong opinion that data in cache is in correct form (satisfies a certain schema) and I'm not sure if there is a certain schema enforcement in all parts of WMCore. Therefore, my argument remains valid since it is better to check the data that it satisfies a schema requirements, e.g. it is in correct data form and you will always get list of strings instead of None data type. And, I doubt that any such check for data-type will cause any harm or penalty but it will certainly protect code from unexpected data.

Reply:
Even though it's a bin script that will be used only a couple of times and be forgotten, I can add that line.